### ISSUE TITLE
TypeError: 'int' object is not iterable

### ISSUE THREAD

**Issue by sukramblak** [jborg/attic#171]
i'm getting this error all the time when i try a second create. 

"client" (debian wheezy) 
* python 3.2 
* attic 0.14 (installed with pip) 

"server" (arch ) 
*  python 3.4
*  attic 0.14 (installed using aur) 

The following complete without errors (and i can verify the existence of the repo on the server) 
* attic init  -v --encryption=keyfile USER@SERVER:Dropbox/Backup/Attic/multi-computer-enc-repos.attic 
* attic create --stats  -v  USER@SERVER:Dropbox/Backup/Attic/multi-computer-enc-repos.attic::sd-sbin /sbin

but a second create .... (example:
* attic create --stats  -v  USER@SERVER:Dropbox/Backup/Attic/multi-computer-enc-repos.attic::sd-etc /etc

results in the following Outout. 
```
/etc
/etc/Bastille
/etc/Bastille/bastille-firewall.cfg
Traceback (most recent call last):
  File "/usr/local/bin/attic", line 3, in <module>
    main()
  File "/usr/local/lib/python3.2/dist-packages/attic/archiver.py", line 727, in main
    exit_code = archiver.run(sys.argv[1:])
  File "/usr/local/lib/python3.2/dist-packages/attic/archiver.py", line 717, in run
    return args.func(args)
  File "/usr/local/lib/python3.2/dist-packages/attic/archiver.py", line 128, in do_create
    self._process(archive, cache, args.excludes, args.exclude_caches, skip_inodes, path, restrict_dev)
  File "/usr/local/lib/python3.2/dist-packages/attic/archiver.py", line 177, in _process
    os.path.join(path, filename), restrict_dev)
  File "/usr/local/lib/python3.2/dist-packages/attic/archiver.py", line 177, in _process
    os.path.join(path, filename), restrict_dev)
  File "/usr/local/lib/python3.2/dist-packages/attic/archiver.py", line 163, in _process
    archive.process_file(path, st, cache)
  File "/usr/local/lib/python3.2/dist-packages/attic/archive.py", line 391, in process_file
    ids = cache.file_known_and_unchanged(path_hash, st)
  File "/usr/local/lib/python3.2/dist-packages/attic/cache.py", line 210, in file_known_and_unchanged
    self._read_files()
  File "/usr/local/lib/python3.2/dist-packages/attic/cache.py", line 84, in _read_files
    for path_hash, item in u:
TypeError: 'int' object is not iterable
```

nearly the same output show up on a **check**
* attic check  USER@SERVER:Dropbox/Backup/Attic/multi-computer-enc-repos.attic
```
Starting repository check...
Repository check complete, no problems found.
Starting archive consistency check...
Analyzing archive sd-sbin (1/1)
Traceback (most recent call last):
  File "/usr/local/bin/attic", line 3, in <module>
    main()
  File "/usr/local/lib/python3.2/dist-packages/attic/archiver.py", line 727, in main
    exit_code = archiver.run(sys.argv[1:])
  File "/usr/local/lib/python3.2/dist-packages/attic/archiver.py", line 717, in run
    return args.func(args)
  File "/usr/local/lib/python3.2/dist-packages/attic/archiver.py", line 84, in do_check
    if not args.repo_only and not ArchiveChecker().check(repository, repair=args.repair):
  File "/usr/local/lib/python3.2/dist-packages/attic/archive.py", line 536, in check
    self.rebuild_refcounts()
  File "/usr/local/lib/python3.2/dist-packages/attic/archive.py", line 686, in rebuild_refcounts
    if b'chunks' in item:
TypeError: argument of type 'int' is not iterable
```

Could you tell please tell me if this a bug (which i was unable to find in the other "issues" on this page) or did i do something wrong during compile or usage?

**Comment by jdchristensen** [comment]
I'm getting the same error when running a backup to a remote host.  But if I run a check on the remote host, there is no error.  This happened shortly after upgrading to attic 0.14 on both machines and also upgrading msgpack (probably from 0.4.2 to 0.4.4, but I can't check right now).  I don't know if it is relevant, but I also found that *both* of the repos I use on the remote host were corrupt shortly after the upgrade;  I repaired them with "attic check --repair", and then "attic check" showed no further errors.  The error below occurred after the repair work had been done.

    # attic create --stats "--exclude-from=/etc/attic/volumes/home.exclude" "user@host:/Backups/attic/home.attic::estrela-home-20150112-04:31" "/home"  "/etc/attic"
    Initializing cache...
    Analyzing archive: stilton-home-20141030-03:31
    Traceback (most recent call last):
      File "/usr/local/bin/attic", line 3, in <module>
        main()
      File "/usr/local/lib/python3.4/dist-packages/attic/archiver.py", line 727, in main
        exit_code = archiver.run(sys.argv[1:])
      File "/usr/local/lib/python3.4/dist-packages/attic/archiver.py", line 717, in run
        return args.func(args)
      File "/usr/local/lib/python3.4/dist-packages/attic/archiver.py", line 100, in do_create
        cache = Cache(repository, key, manifest)
      File "/usr/local/lib/python3.4/dist-packages/attic/cache.py", line 33, in __init__
        self.sync()
      File "/usr/local/lib/python3.4/dist-packages/attic/cache.py", line 168, in sync
        if b'chunks' in item:
    TypeError: argument of type 'int' is not iterable

**Comment by jdchristensen** [comment]
I got this error again in a completely independent set-up, a different client machine connecting to a different backup server.  Both were recently upgraded to attic 0.14 and msgpack 0.4.4.  This time the check command on the repo (run locally) showed no errors.

Both this case and the one I mentioned in my previous comment share some common features:  the backup is to a remote machine over ssh; multiple clients back-up to the same repo; recent upgrades.

The backtrace is slightly different, showing a keyerror and then another backtrace after that:

    attic create --stats "--exclude-from=/etc/attic/volumes/system.exclude" "bu@jdc:/Backups/attic/system.attic::estrela-system-20150114-09:22" "/"
    Initializing cache...
    Analyzing archive: jdc-system-20141101-00:14
    Traceback (most recent call last):
      File "/usr/local/lib/python3.4/dist-packages/attic/remote.py", line 300, in get_many
        yield self.load_object(*self.index[key])
      File "hashindex.pyx", line 96, in attic.hashindex.NSIndex.__getitem__ (attic/hashindex.c:2248)
    KeyError
    
    During handling of the above exception, another exception occurred:
    Traceback (most recent call last):
      File "/usr/local/bin/attic", line 3, in <module>
        main()
      File "/usr/local/lib/python3.4/dist-packages/attic/archiver.py", line 727, in main
        exit_code = archiver.run(sys.argv[1:])
      File "/usr/local/lib/python3.4/dist-packages/attic/archiver.py", line 717, in run
        return args.func(args)
      File "/usr/local/lib/python3.4/dist-packages/attic/archiver.py", line 100, in do_create
        cache = Cache(repository, key, manifest)
      File "/usr/local/lib/python3.4/dist-packages/attic/cache.py", line 33, in __init__
        self.sync()
      File "/usr/local/lib/python3.4/dist-packages/attic/cache.py", line 163, in sync
        for key, chunk in zip(archive[b'items'], repository.get_many(archive[b'items'])):
      File "/usr/local/lib/python3.4/dist-packages/attic/remote.py", line 302, in get_many
        for key_, data in repository_iterator:
      File "/usr/local/lib/python3.4/dist-packages/attic/remote.py", line 232, in get_many
        for resp in self.call_many('get', [(id_,) for id_ in ids], is_preloaded=is_preloaded):
      File "/usr/local/lib/python3.4/dist-packages/attic/remote.py", line 175, in call_many
        for type, msgid, error, res in self.unpacker:
    TypeError: 'int' object is not iterable

**Comment by t-szczyrba** [comment]
I get similar error issuing info command in one example repo:
# attic info rootfs::root-2015-01-15 

-- cut --
Traceback (most recent call last):
  File "/usr/local/bin/attic", line 4, in <module>
    __import__('pkg_resources').run_script('Attic==0.14', 'attic')
  File "/usr/local/lib/python3.2/dist-packages/setuptools-11.3.1-py3.2.egg/pkg_r
esources/__init__.py", line 696, in run_script
  File "/usr/local/lib/python3.2/dist-packages/setuptools-11.3.1-py3.2.egg/pkg_r
esources/__init__.py", line 1614, in run_script
  File "/usr/local/lib/python3.2/dist-packages/Attic-0.14-py3.2-linux-i686.egg/EGG-INFO/scripts/attic", line 3, in <module>
    main()
  File "/usr/local/lib/python3.2/dist-packages/Attic-0.14-py3.2-linux-i686.egg/attic/archiver.py", line 727, in main
    exit_code = archiver.run(sys.argv[1:])
  File "/usr/local/lib/python3.2/dist-packages/Attic-0.14-py3.2-linux-i686.egg/attic/archiver.py", line 717, in run
    return args.func(args)
  File "/usr/local/lib/python3.2/dist-packages/Attic-0.14-py3.2-linux-i686.egg/attic/archiver.py", line 307, in do_info
    stats = archive.calc_stats(cache)
  File "/usr/local/lib/python3.2/dist-packages/Attic-0.14-py3.2-linux-i686.egg/attic/archive.py", line 221, in calc_stats
    if b'chunks' in item:
TypeError: argument of type 'int' is not iterable
---

I have checked the behaviour twice (freshly creating archive) with 0.13 and 0.14
Python 3.2.3, ubuntu 12.04.
What more: after upgrade with such an archive to python 3.4.1 and testing with attic 0.14 info command was issued succesfully, but the archive was corrupted.
Creating the archive once again with fs not heavily modified made everything running fine: attic reports no problems with info as well as check command.

Running attic 0.14 in opensuse 13.2 ( Python 3.4.1 ) seems much more robust  than with python 3.2 - maybe running attic with python 3.2 is not really a good idea?

**Comment by jdchristensen** [comment]
I was able to work around this problem by downgrading msgpack-python from 0.4.4 to 0.4.2.

I also installed g++, which means that msgpack-python doesn't fall back to its pure python version.

**Comment by rbu** [comment]
@jdchristensen Thank you for the hint. I have debugged this down to a reproducer that I'll share with with msgpack folks. I have opened https://github.com/msgpack/msgpack-python/issues/124

**Comment by jdchristensen** [comment]
@rbu:  so does the compiled version of msgpack 0.4.4 work for you with attic?

In the meantime, could attic detect that it is running with a buggy version and refuse to run?

**Comment by rbu** [comment]
I could not reproduce this issue with the compiled version of msgpack 0.4.4 as well as 0.4.5. At least for me, this issue only occurs when *reading* data. So if I create an archive with Pure-Python msgpack that fails ```attic check``` with that same version, once I install a C-compiled msgpack, the "failing" archive now passes the check.
I'm not sure if that totally excludes the case of data corruption when having used a Pure-Python msgpack in the past, but at least it was possible for me to resurrect the data.

**Comment by jdchristensen** [comment]
Great!  Maybe there's a way to change attic so it also works with the pure python version?  E.g. instead of reading a msgpack iterator more than once, it could save the data it reads and reuse it?  Or something like that.

**Comment by rbu** [comment]
msgpack upstream confirmed the issue and they are working on a fix. Apparently this problem was introduced after 0.4.2, so for that version both Pure-Python and C builds are fine.

**Comment by jborg** [comment]
Seems like the msgpack guys have fixes this (msgpack/msgpack-python#124) now so I'm closing this issue now.

**Comment by ThomasWaldmann** [comment]
See PR #230.
